### PR TITLE
fix(rust, python): return appropriate dtypes for temporal `mean` and `median`

### LINKED ISSUE #10232: Median on duration Series returns a datetime rather than a timedelta
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
from datetime import timedelta
import polars as pl
s = pl.Series("s", [timedelta(0), timedelta(1)])
s.describe()
```

Output:
```
shape: (5, 2)
┌────────────┬─────────────────────┐
│ statistic  ┆ value               │
│ ---        ┆ ---                 │
│ str        ┆ str                 │
╞════════════╪═════════════════════╡
│ count      ┆ 2                   │
│ null_count ┆ 0                   │
│ min        ┆ 0:00:00             │
│ max        ┆ 1 day, 0:00:00      │
│ median     ┆ 1970-01-01 12:00:00 │
└────────────┴─────────────────────┘
```

### Issue description

The median shows up as a datetime in .describe() of a Series.  But in a dataframe, the result is correctly a timedelta/duration.

### Expected behavior

Expected:
```
shape: (5, 2)
┌────────────┬─────────────────────┐
│ statistic  ┆ value               │
│ ---        ┆ ---                 │
│ str        ┆ str                 │
╞════════════╪═════════════════════╡
│ count      ┆ 2                   │
│ null_count ┆ 0                   │
│ min        ┆ 0:00:00             │
│ max        ┆ 1 day, 0:00:00      │
│ median     ┆ 12:00:00            │
└────────────┴─────────────────────┘
```

### Installed versions

<details>

```
--------Version info---------
Polars:              0.18.11
Index type:          UInt32
Platform:            Linux-3.10.0-1160.92.1.el7.x86_64-x86_64-with-glibc2.17
Python:              3.9.17 (main, Jun  8 2023, 11:31:03) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
matplotlib:          <not installed>
numpy:               <not installed>
pandas:              <not installed>
pyarrow:             <not installed>
pydantic:            <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by mcrumiller**
**This is a breaking change**

Resolves #10232. Calls to `mean` and `median` now have the following return dtypes:

Dtype | `Series.mean()` output
------|--------
`Time` | `datetime.time`
`Date` | `datetime.datetime("ms")`
`Datetime(tu, tz)` | `datetime.datetime(tu, tz)`
`Duration(tu)` | `datetime.timedelta(tu)`

`pl.Date` returns a `pl.Datetime` because this case is analagous to integers: whereas the mean of 1 and 2 is 0.5, now the mean of `2023-01-01` and `2023-01-02` is `2023-01-02 12:00:00`.

---

@MarcoGorelli This is the first of three PRs related to temporals. I think I addressed your issues from my prior too-big PR, along with additional improvements:

1. Moved the unit conversion in-line, no more const up front.
2. Add tests for both regular `mean`/`median`and `dt` namespace.
3. Changed return type typing.
4. Included `pl.Time` dtype
5. Implemented for `group_by.mean`/`group_by.median`

### CODE REVIEW

File: crates/polars-core/src/frame/group_by/aggregations/dispatch.rs [hunk 1]
```diff
@@ -109,19 +111,29 @@ impl Series {
         use DataType::*;
 
         match self.dtype() {
+            Boolean => self.cast(&Float64).unwrap().agg_mean(groups),
```
Comment (MarcoGorelli) [inline crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:114]: should this be `agg_median`?

can you also add a test case for this please, asserting that for both
```python
In [12]: pl.Series([True, True, False]).median()
Out[12]: 1.0

In [13]: pl.Series([True, True, False]).to_frame('a').group_by(1).agg(pl.col('a').median())
Out[13]: 
shape: (1, 2)
┌─────────┬──────────┐
│ literal ┆ a        │
│ ---     ┆ ---      │
│ i32     ┆ f64      │
╞═════════╪══════════╡
│ 1       ┆ 0.666667 │
└─────────┴──────────┘
```
the output should be `1.0`?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: crates/polars-core/src/frame/group_by/aggregations/dispatch.rs (hunk 1 above)
Comment (mcrumiller) [reply crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:114]: Yes, good catch, thank you.

**Review by MarcoGorelli — COMMENTED** [review]
Awesome work!

Just got some nitpicks but - barring requests from others - I think looks good to me

Just to check my understanding - in the follow-up, the plan is to deprecate `.dt.mean` / `.dt.median` in favour of just `.mean` and `.median`, right?

I think the rest is solid - will sleep on it and do a final check before approving

File: crates/polars-core/src/frame/group_by/aggregations/dispatch.rs [hunk 2]
```diff
@@ -109,19 +111,29 @@ impl Series {
...
-                }
+            dt if dt.is_numeric() => {
+                apply_method_physical_integer!(self, agg_median, groups)
+            },
+            #[cfg(feature = "dtype-date")]
+            Date => {
+                let s = self.cast(&Int64).unwrap() * (US_IN_DAY as f64);
+                // agg_median returns Float64
+                let out = s.agg_median(groups);
+                // cast back to Int64 and then to logical duration type
```
[the remark this answers is not in the document]
Comment (mcrumiller) [reply crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:125]: Yep, definitely. Will fix once I'm out of my meeting....

File: py-polars/tests/unit/operations/test_group_by.py [hunk 3]
```diff
@@ -121,6 +124,204 @@ def test_group_by_args() -> None:
     assert df.group_by("a").agg(q="b", r="c").columns == ["a", "q", "r"]
 
 
+@pytest.mark.parametrize(
+    ("input", "expected", "input_dtype", "output_dtype"),
+    [
+        # fmt: off
```
[the remark this answers is not in the document]
Comment (mcrumiller) [reply py-polars/tests/unit/operations/test_group_by.py]: Yeah sorry, had that on during dev so I could actually read the input (88 chars is not enough for these parametrized inputs).

**Comment by mcrumiller** [comment]
@MarcoGorelli I think I hit all of your nits. I have a few followups:

1. deprecate `dt.mean`, `dt.median`, `dt.min`, and `dt.max` as those provide nothing of value of `series.*`.
2. Implement `std` and `var` for temporal dtypes and address #12896. I already have a working implementation for that. Could  split into two PRs but they are very closely related.
3. (maybe) implement `std` and `var` for `group_by`, which is currently missing:
    ```python
    import polars as pl
    
    pl.DataFrame({
        "key": [1, 1, 2, 2],
        "value": [1, 2, 3, 4],
    }).group_by("key").std()
    ```
    ```
    AttributeError: 'GroupBy' object has no attribute 'std'
    ```

**Comment by mcrumiller** [comment]
The error seems to be related to a streaming file, I don't think it's related to this PR.

**Review by MarcoGorelli — APPROVED** [review]
Nice work, looks good to me! Tests look very thorough

Quick demo to summarise the changes:

- on the latest release:
  ```python
  In [1]: pl.Series([date(2020, 1, 1), date(2020, 1, 2)]).mean()
  Out[1]: 18262.5

  In [2]: pl.Series([date(2020, 1, 1), date(2020, 1, 2)]).dt.mean()
  Out[2]: datetime.date(2020, 1, 1)

  In [3]: pl.Series([date(2020, 1, 1), date(2020, 1, 2)]).to_frame('a').select(pl.col('a').mean())
  Out[3]:
  shape: (1, 1)
  ┌──────┐
  │ a    │
  │ ---  │
  │ date │
  ╞══════╡
  │ null │
  └──────┘
  ```

- here:
  ```python
  In [1]: pl.Series([date(2020, 1, 1), date(2020, 1, 2)]).mean()
  Out[1]: datetime.datetime(2020, 1, 1, 12, 0)

  In [2]: pl.Series([date(2020, 1, 1), date(2020, 1, 2)]).dt.mean()  # todo: deprecate in separate PR
  Out[2]: datetime.datetime(2020, 1, 1, 12, 0)

  In [3]: pl.Series([date(2020, 1, 1), date(2020, 1, 2)]).to_frame('a').select(pl.col('a').mean())
  Out[3]:
  shape: (1, 1)
  ┌─────────────────────┐
  │ a                   │
  │ ---                 │
  │ datetime[μs]        │
  ╞═════════════════════╡
  │ 2020-01-01 12:00:00 │
  └─────────────────────┘
  ```
, and analogously for median, as well as for Datetime/Duration/Time

**Comment by mcrumiller** [comment]
Thanks Marco!

**Comment by ritchie46** [comment]
Can you rebase @mcrumiller?

**Comment by mcrumiller** [comment]
> Can you rebase @mcrumiller?

@ritchie46 done.

**Comment by mcrumiller** [comment]
@ritchie46 I think this one is finally good to go.

**Comment by ritchie46** [comment]
This one needs quite a rebase. :/

**Comment by mcrumiller** [comment]
Yep, working on it.

File: crates/polars-core/src/series/implementations/datetime.rs [hunk 4]
```diff
@@ -356,20 +356,32 @@ impl SeriesTrait for SeriesWrap<DatetimeChunked> {
...
+
     fn quantile_as_series(
         &self,
         _quantile: f64,
         _interpol: QuantileInterpolOptions,
     ) -> PolarsResult<Series> {
-        Ok(Int32Chunked::full_null(self.name(), 1)
-            .cast(self.dtype())
-            .unwrap())
+        Int32Chunked::full_null(self.name(), 1).cast(self.dtype())
```
Comment (mcrumiller) [inline crates/polars-core/src/series/implementations/datetime.rs:384]: Not part of my PR, but was in an adjacent function and small enough not to warrant another PR. No need to unwrap/rewrap.

**Comment by mcrumiller** [comment]
Probably going to get a docstring error re: `df.describe`.

**Comment by mcrumiller** [comment]
@ritchie46 good to go.

**Comment by MarcoGorelli** [comment]
there's another conflict

**Comment by mcrumiller** [comment]
Thanks for the heads up @MarcoGorelli (can we configure our CI to send out an email when a PR becomes stale?).

**Comment by mcrumiller** [comment]
@ritchie46 and @MarcoGorelli I've fixed the [minor] conflict.

File: crates/polars-core/src/frame/group_by/aggregations/dispatch.rs [hunk 5]
```diff
@@ -109,19 +111,29 @@ impl Series {
...
-                    s.cast(physical_type).unwrap().cast(dt).unwrap()
-                } else {
-                    s
-                }
+            dt if dt.is_numeric() => {
+                apply_method_physical_integer!(self, agg_median, groups)
+            },
+            #[cfg(feature = "dtype-date")]
+            Date => {
+                let s = self.cast(&Int64).unwrap() * (US_IN_DAY as f64);
```
Comment (ritchie46) [inline crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:122]: Why the multiplication here? We can just use the physical type right?

And shouldn't this return `Date`?

File: crates/polars-core/src/series/implementations/dates_time.rs [hunk 6]
```diff
@@ -348,10 +350,43 @@ macro_rules! impl_dyn_series {
...
             }
+
             fn min_as_series(&self) -> PolarsResult<Series> {
                 Ok(self.0.min_as_series().$into_logical())
             }
 
+            fn median_as_series(&self) -> PolarsResult<Series> {
+                match self.dtype() {
+                    #[cfg(feature = "dtype-date")]
+                    DataType::Date => {
```
Comment (ritchie46) [inline crates/polars-core/src/series/implementations/dates_time.rs:361]: idem

File: crates/polars-core/src/series/implementations/datetime.rs [hunk 7]
```diff
@@ -356,20 +356,32 @@ impl SeriesTrait for SeriesWrap<DatetimeChunked> {
...
             .0
             .min_as_series()
             .into_datetime(self.0.time_unit(), self.0.time_zone().clone()))
     }
+
+    fn median_as_series(&self) -> PolarsResult<Series> {
+        Series::new(self.name(), &[self.median().map(|v| v as i64)]).cast(self.dtype())
+    }
+
+    fn var_as_series(&self, _ddof: u8) -> PolarsResult<Series> {
```
Comment (ritchie46) [inline crates/polars-core/src/series/implementations/datetime.rs:371]: This should be removed because we want to raise if not supported for a dtype.

File: crates/polars-core/src/series/implementations/datetime.rs [hunk 8]
```diff
@@ -356,20 +356,32 @@ impl SeriesTrait for SeriesWrap<DatetimeChunked> {
...
+        Int32Chunked::full_null(self.name(), 1).cast(self.dtype())
+    }
+
+    fn std_as_series(&self, _ddof: u8) -> PolarsResult<Series> {
```
Comment (ritchie46) [inline crates/polars-core/src/series/implementations/datetime.rs:375]: This should be removed because we want to raise if not supported for a dtype.

File: crates/polars-core/src/series/implementations/datetime.rs [hunk 9]
```diff
@@ -356,20 +356,32 @@ impl SeriesTrait for SeriesWrap<DatetimeChunked> {
...
+        Int32Chunked::full_null(self.name(), 1).cast(self.dtype())
+    }
+
     fn quantile_as_series(
```
Comment (ritchie46) [inline crates/polars-core/src/series/implementations/datetime.rs:379]: This should be removed because we want to raise if not supported for a dtype.

**Review by ritchie46 — CHANGES_REQUESTED** [review]

File: crates/polars-core/src/frame/group_by/aggregations/dispatch.rs (hunk 5 above)
Comment (MarcoGorelli) [reply crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:122]: on the return type, there was a bit of discussion here https://github.com/pola-rs/polars/pull/12670#pullrequestreview-1749268415

I think it can make sense to upcast `Date` to `Datetime('us')`, as they're limited to the same range of dates (+/- 262,000 years since unix epoch)

Comment (mcrumiller) [reply crates/polars-core/src/frame/group_by/aggregations/dispatch.rs:122]: @ritchie46 it's in the OP, the same way the mean/median of ints is a float, the mean/median of dates is now a datetime.

File: crates/polars-core/src/series/implementations/datetime.rs (hunk 7 above)
Comment (mcrumiller) [reply crates/polars-core/src/series/implementations/datetime.rs:371]: I can add that to this PR, although I was going to implement this in my next PR anyway.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Seeing as this is open, there's one more thing I've thought of to help future-proof things

Currently, the way you've done it, there's no risk of overflow anyway because of Chrono's limitation to years 262,000 +/- unix epoch. But if Chrono we to lift that limitation, or if Polars were to move away from Chrono, then `Datetime('ms')` would still avoid any risk of overflows:
```
In [25]: np.datetime64('1970-01-01') + np.timedelta64(2**31-1, 'D')
Out[25]: numpy.datetime64('5881580-07-11')

In [26]: np.datetime64('1970-01-01') + np.timedelta64(2**63-1, 'us')
Out[26]: numpy.datetime64('294247-01-10T04:00:54.775807')

In [27]: np.datetime64('1970-01-01') + np.timedelta64(2**63-1, 'ms')
Out[27]: numpy.datetime64('292278994-08-17T07:12:55.807')
```
So this would help future-proof things, and avoid risking angering paleoclimatologists if/when that day comes

**Comment by mcrumiller** [comment]
@MarcoGorelli, that seems a bit of an excessive future-proofing. Surely if chrono lifts that limitation, we can address the issue then? `us` are the default datetime units throughout polars and it seems fitting to return the `us` type.

**Comment by MarcoGorelli** [comment]
> we can address the issue then? 

how? What would you return as the average of `5881580-01-01` and `5881580-01-02`?

**Comment by mcrumiller** [comment]
Well, my thought was maybe *then* we could decide to return `ms` if we were maybe above that limit. But ok, let's future proof now. Should we always return `ms` for mean/median, or check for overflow and then return `ms` depending on the result?

**Comment by MarcoGorelli** [comment]
The output dtype needs to be predictable regardless of the values (that's something to check actually - if you have lazyframe and take the mean of a temporal column, is the output schema correct?), so `'ms'` consistently would be better

**Comment by mcrumiller** [comment]
Prior to this PR, date returned a `pl.Date` for mean/median and datetime returned the same units.

I'm fine using `ms` here, I'll get that in in a bit.

**Comment by mcrumiller** [comment]
@MarcoGorelli I've updated to return `ms` and updated the unit tests to check for `ms` when the input dtype is a date. I updated the OP above too.

**Comment by mcrumiller** [comment]
Got a possible import speed regression...unrelated to PR but maybe is an issue?

**Comment by mcrumiller** [comment]
Are there any more concerns? I'm pretty eager to get this one finished since I have a few followups that should hopefully go a bit more smoothly.

**Review by MarcoGorelli — COMMENTED** [review]
> Prior to this PR, date returned a `pl.Date` for mean/median and datetime returned the same units.

One issue still present is that the lazy schema doesn't reflect the new output dtype:
```python
df = pl.LazyFrame({
    'a': [date(2020, 1, 1), date(2020, 1, 2)],
    'b': [datetime(2020, 1, 1), datetime(2020, 1, 2)],
    'c': [timedelta(1), timedelta(2)],
})
print(df.select(pl.all().mean()).schema)
print(df.select(pl.all().mean()).collect().schema)
```
```
OrderedDict([('a', Date), ('b', Datetime(time_unit='us', time_zone=None)), ('c', Duration(time_unit='us'))])
OrderedDict([('a', Datetime(time_unit='ms', time_zone=None)), ('b', Datetime(time_unit='us', time_zone=None)), ('c', Duration(time_unit='us'))])
```

---

The other thing is that this PR now introduces a breaking change for `Series.std` for temporal dtypes - before it would return `None`, whereas now it raises?
This looks completely orthogonal to this PR, right? If so, I'd suggest breaking that off into a separate PR (smaller, self-contained PRs that only do a single thing are faster to review and merge)

**Comment by mcrumiller** [comment]
Thanks @MarcoGorelli . The removal of `std` and `var` was at Ritchie's request since unsupported functions now raise instead of returning `None` and I think those were missed. I can move that to a separate PR of course.

I'll check out the LazyFrame issue later today (on the road for a bit for the holidays).

**Comment by mcrumiller** [comment]
@MarcoGorelli I'm almost there. I got the schema working for `ldf`, but there is one final test that fails: `group_by().mean().collect(streaming=True)` no longer works and returns 1970-01-01:

```python
from datetime import date
import polars as pl

ldf = pl.LazyFrame({
    "key": [1, 1],
    "a": [date(2020, 1, 1), date(2020, 1, 2)],
})

ldf.group_by("key").mean().collect()               # works
ldf.group_by("key").mean().collect(streaming=True) # returns 1970-01-01 00:00:18.262
```

I've been trying to find the code path here but I could use a bit of help. Thanks!

File: crates/polars-lazy/src/physical_plan/streaming/convert_alp.rs [hunk 10]
```diff
@@ -442,6 +442,10 @@ pub(crate) fn insert_streaming_nodes(
                         .all(|dt| allowed_dtype(dt, string_cache))
                 };
 
+                let can_valid_agg = valid_agg();
+                let can_valid_key = valid_key();
+                let can_valid_types = valid_types();
```
Comment (MarcoGorelli) [inline crates/polars-lazy/src/physical_plan/streaming/convert_alp.rs:447]: looks unrelated?

**Comment by mcrumiller** [comment]
@MarcoGorelli schema is fixed, and `collect(streaming=True)` is now fixed as well. I think all is good, can you take a look?

**Review by MarcoGorelli — COMMENTED** [review]
thanks! will check in the week, looks like something unrelated has crept in though?

File: crates/polars-lazy/src/physical_plan/streaming/convert_alp.rs (hunk 10 above)
Comment (mcrumiller) [reply crates/polars-lazy/src/physical_plan/streaming/convert_alp.rs]: Quick catch! Thanks, fixed.

**Review by MarcoGorelli — COMMENTED** [review]
Sorry to be a pain, but I think this is still doing a lot in the same PR, and it's easy to miss little things when reviewing 19 files

I think you're doing 3 things here:
1. `s = pl.Series([datetime(2020, 1, 1)]); s.mean()` now returns a `datetime`, instead of a float
2. `s.to_frame('a').select(pl.col('a').mean())` now returns a 1-row `datetime` column, instead of a 1-row column with `null`
3. `s = pl.Series([date(2020, 1, 1)]); s.mean()` now returns a `datetime`

I think 1 and 2 probably need to go together, but 3 can be done separately, right?

If so, could we please start with just 1 and 2, and do 3 after?

**Comment by mcrumiller** [comment]
Hey @MarcoGorelli, not a pain at all, no worries--I'd rather make this easier for you. I can split this up into a series of more atomic PRs. This one has become a bit messy anyway.

**Comment by mcrumiller** [comment]
Closing in favor of more atomic PRs.
